Flatpak Build Support

iAmSomeone2 opened the pull request:
Based on issue #366, I've begun work on a Flatpak build config.

## Testing this PR

### Prerequisites

1. Ensure both `flatpak` and `flatpak-builder` are installed on your system.
2. Ensure that, at least, the Flathub repo is enabled by following the instructions for your OS found here: https://flathub.org/setup
3. Use `flatpak` to install the KDE platform and SDK:
   ```sh
   flatpak install org.kde.Platform//6.8 org.kde.Sdk//6.8
   ```

### Building

From the project root run:
 ```sh
 flatpak-builder --user --install flatpak-build flatpak/pizza.isle.Isle.json
 ```
and wait for the build to complete.

If successful, the Flatpak will be installed for the current user and can be found in your applications menu.

**NOTE**: Making a change and running the above command again will install the new version over the old one. You don't have to remove the old Flatpak yourself.

### On First Run

1. Load a physical CD or mount an ISO of the game files.
2. Locate the "Isle Decomp" entry in your applications menu and right-click it to open the context menu. Click the "Configure" action to launch the config app.
3. Set the "Disk" and/or "CD" paths to paths in the mounted disk (image). For example:
   > Disk: `/run/media/<username>/LEGO_ISLANDI/DATA/disk`
   > CD: `/run/media/<username>/LEGO_ISLANDI`
4. **IMPORTANT**: You must not use the file picker buttons for this. The sandbox settings will need to be tuned more to make that work as expected.
5. Adjust any other settings as you wish and save the config.

### Running the App & Testing the Packaging

Once configured, "Isle Decomp" can be launched from the app menu the same as any other app. If your paths were set correctly, it should run like normal.

Metadata can be viewed in your OS's software center. For example, Ubuntu and Fedora Workstation users can open GNOME Software, find the app under the "Installed" tab, and open its entry to see how a public listing would look to an end user.

AJenbo commented on flatpak/pizza.isle.Isle.json:
```diff
@@ -0,0 +1,68 @@
+{
+    "id": "pizza.isle.Isle",
+
+    "runtime": "org.kde.Platform",
```
Use the generic runtime instead of the kde one, you can use this as reference:  https://github.com/flathub/org.diasurgical.DevilutionX/blob/master/org.diasurgical.DevilutionX.yml

iAmSomeone2 replied:
I took a look at your example, and I can see where you're coming from.

If we only had to package the "isle" binary, then the Freedesktop platform runtime would be enough; however, the "config" binary requires Qt 6 and [Portal](https://docs.flatpak.org/en/latest/sandbox-permissions.html#portals) support, and the best way to accomplish that is by using the KDE runtime. While making everything work under the Freedesktop runtime is possible, it would involve a significant increase in packaging work and ongoing maintenance that has no benefit to the end user or improved portability.

Flatpak runtimes are distro and DE-agnostic (ex. KDE works on GNOME and vice versa). They are also not bundled with the packages themselves. Instead, the package contains some metadata which tells the installer which runtime it uses and what version. Flatpak ensures that there is only ever one copy of a given runtime installed at a time. If your app's requested runtime has already been installed by another app, then nothing there changes. Otherwise, Flatpak installs the runtime and makes it available for any app that requests it.

This system ends up being like a hybrid between containers and traditional shared libraries; apps are sandboxed with tight permission controls and have access to the exact libraries and runtime environment they need while also reducing duplication of the most common libraries for desktop apps. Or from purely a packaging perspective, it's like having bundles of versioned shared libraries that you can guarantee will be the same and validated to work on end user systems so long as they can run a supported version of Flatpak.

As an example setup, let's say that I already have the [Dolphin Emu Flatpak](https://github.com/dolphin-emu/dolphin/blob/master/Flatpak/org.DolphinEmu.dolphin-emu.yml) installed. As part of that installation, Flatpak also verified that the KDE 6.8 runtime was installed on my system. Now, if I bump our requested version down to 6.8 and install the _Isle Portable_ package, Flatpak will see that our package also wants KDE 6.8, so nothing new needs to be downloaded. Therefore, our actual package size is only what we need on top of the runtime: a couple small binaries, some metadata files, and desktop icons. In all, it currently comes out to just 6.2MiB for the x86_64 build!

AJenbo reviewed (commented):
Bring able to install it as a dev is not.particularly interesting, please make sure it's build by the ci so it can be included in releases

AJenbo commented on BUILD_DEPS.md:
```diff
@@ -0,0 +1,25 @@
+# Platform-Specific Build Dependencies
```
Should this be moved to docs/building.md ?

foxtacles replied:
Sounds good

AJenbo commented on flatpak/desktop/pizza.isle.Isle.metainfo.xml:
```diff
@@ -0,0 +1,51 @@
+
+  <keywords>
+    <keyword translate="no">LEGO</keyword>
+    <keyword translate="no">lego</keyword>
```
except for error messages there is no text to translate and it works with all the translated assets.

AJenbo commented on packaging/linux/isle_decomp.metainfo.xml.in:
```diff
@@ -0,0 +1,86 @@
+    <keyword>LEGO Island</keyword>
+    <!-- Translations sourced from https://www.pcgamingwiki.com/wiki/Lego_Island#Localizations  -->
+    <keyword xml:lang="fr">Aventures sur L'île LEGO</keyword>
+    <keyword xml:lang="de">Abenteuer auf der LEGO Insel</keyword>
```
```suggestion
    <keyword xml:lang="da">Titled Panik på LEGO Øen</keyword>
    <keyword xml:lang="de">Abenteuer auf der LEGO Insel</keyword>
```

AJenbo commented on packaging/linux/isle_decomp.desktop.in:
```diff
@@ -0,0 +1,32 @@
+Categories=Game;KidsGame;AdventureGame;Qt
+
+Keywords=LEGO;lego;LEGO Island
+Keywords[es]=LEGO;lego;La Isla LEGO
+Keywords[fr]=LEGO;lego;Aventures sur L'île LEGO
+Keywords[it]=LEGO;lego;Isola LEGO
+Keywords[de]=LEGO;lego;Abenteuer auf der LEGO Insel
+Keywords[jp]=LEGO;lego;レゴアイランドの大冒険
+Keywords[pt_BR]=LEGO;lego;A Ilha LEGO
```
```suggestion
Keywords[da]=LEGO;lego;Titled Panik på LEGO Øen
Keywords[de]=LEGO;lego;Abenteuer auf der LEGO Insel
Keywords[es]=LEGO;lego;La Isla LEGO
Keywords[fr]=LEGO;lego;Aventures sur L'île LEGO
Keywords[it]=LEGO;lego;Isola LEGO
Keywords[jp]=LEGO;lego;レゴアイランドの大冒険
Keywords[pt_BR]=LEGO;lego;A Ilha LEGO
```

AJenbo commented on packaging/linux/isle_decomp.metainfo.xml.in:
```diff
@@ -0,0 +1,87 @@
+
+    <keyword>LEGO Island</keyword>
+    <!-- Translations sourced from https://www.pcgamingwiki.com/wiki/Lego_Island#Localizations  -->
+    <keyword xml:lang="fr">Aventures sur L'île LEGO</keyword>
+    <keyword xml:lang="de">Abenteuer auf der LEGO Insel</keyword>
+    <keyword xml:lang="da">Titled Panik på LEGO Øen</keyword>
+    <keyword xml:lang="it">Isola LEGO</keyword>
+    <keyword xml:lang="es">La Isla LEGO</keyword>
+    <keyword xml:lang="jp">レゴアイランドの大冒険</keyword>
+    <keyword xml:lang="pt-BR">A Ilha LEGO</keyword>
```
```suggestion
    <keyword xml:lang="da">Titled Panik på LEGO Øen</keyword>
    <keyword xml:lang="de">Abenteuer auf der LEGO Insel</keyword>
    <keyword xml:lang="es">La Isla LEGO</keyword>
    <keyword xml:lang="fr">Aventures sur L'île LEGO</keyword>
    <keyword xml:lang="it">Isola LEGO</keyword>
    <keyword xml:lang="jp">レゴアイランドの大冒険</keyword>
    <keyword xml:lang="pt-BR">A Ilha LEGO</keyword>
```

foxtacles commented on .github/workflows/release.yml:
```diff
@@ -91,15 +92,53 @@ jobs:
+    runs-on: ${{ matrix.os }}
+
+    env:
+      APP_ID: pizza.isle.Isle
```
I don't know the nature of the app IDs, but we generally use isledecomp as the org, so I imagine something like isledecomp.isle.Isle is better

iAmSomeone2 replied:
Sounds good! The full ID will look like `org.legoisland.Isle`, then.

foxtacles commented on packaging/linux/isle_decomp.desktop.in:
```diff
@@ -0,0 +1,33 @@
+[Desktop Entry]
```
file name should be isledecomp, the others too

iAmSomeone2 replied:
Done! It'll be in my next commit.

madebr commented on CMakeLists.txt:
```diff
@@ -2,6 +2,15 @@ cmake_minimum_required(VERSION 3.25...4.0 FATAL_ERROR)
+if (CONFIGURE_FLATPAK)
+  # Exit early since the remaining configuration will otherwise fail
+  return()
+endif()
```
I think the `isledecomp.manifest.json.ini` is needlessly made configurable.
Only the install xml manifest should be configured. no?
I also don't like how this option skips a large part of the cmake script. That's a no-no imho.
Flatpak is also nix-only at the moment, so I think the `linux` subdirectory is not needed?

Can't the cmake script install all files to its appropriate locations? Such that `install_desktop_assets.sh` is not needed anymore? 
e.g.
```cmake
option(CONFIGURE_FLATPAK "Configure Flatpak data" OFF)
if(CONFIGURE_FLATPAK)
  set(APP_ID "org.legoisland.Isle")
  set(APP_NAME "Isle Portable")
  set(APP_SUMMARY "Portable version of the LEGO Island Decompilation Project")
  set(APP_SPDX "LGPL-3.0-or-later")
  set(QT_VERSION 6.8)
  string(TIMESTAMP BUILD_DATE UTC)
  execute_process(
    COMMAND git log -1 --format=%h
    WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
    OUTPUT_VARIABLE GIT_HASH
    OUTPUT_STRIP_TRAILING_WHITESPACE
  )
  if(LINUX)
    configure_file(isledecomp.desktop.in "${APP_ID}.desktop" @ONLY)
    configure_file(isledecomp.metainfo.xml.in "${APP_ID}.metainfo.xml" @ONLY)
    install(FILES "icons/isle.xvg" DESTINATION "scalable/apps")
    install(FILES "${CMAKE_BINARY_DIR}/${APP_ID}.desktop" DESTINATION "share/applications")
    install(FILES "${CMAKE_BINARY_DIR}/${APP_ID}.desktop" DESTINATION "share/applications")
  endif()
endif()
```

iAmSomeone2 replied:
You're definitely right about adding the install directives to the CMake scripts. It's been a good 4 years since I've used CMake (or C++) regularly, so it's easy to forget what it can do.

For your highlighted lines, ending the CMake script early in this situation is very intentional. Flatpak builds happen in a container-like environment and do not use any libraries from the host OS; however, these metadata files must be prepared before the build starts. The full configure step does run inside of the Flatpak build environment. 

The `CONFIGURE_FLATPAK` option is intended to run the configure step just long enough to prepare the metadata to avoid having to download libraries that will simply be ignored just to make CMake happy. Obviously this isn't an issue for local dev, but it just wastes time and bandwidth on the pipeline runners. Of course, this does remind me that `custom_target` exists and might be the solution we're looking for here. Maybe configuring the metadata files could be its own target?

iAmSomeone2 replied:
The Flatpak manifest is static again; however, it had to be renamed back to the `$APP_ID.json` pattern since CMake no longer touches it.

iAmSomeone2 commented:
You guys can ignore my pushes for a bit. I'm debugging the release pipeline.

AJenbo commented on packaging/linux/isledecomp.desktop.in:
```diff
@@ -0,0 +1,33 @@
+Type=Application
+Categories=Game;KidsGame;AdventureGame;Qt
+
+Keywords=LEGO;lego;LEGO Island
+Keywords[da]=LEGO;lego;Titled Panik på LEGO Øen
+Keywords[de]=LEGO;lego;Abenteuer auf der LEGO Insel
+Keywords[es]=LEGO;lego;La Isla LEGO
+Keywords[fr]=LEGO;lego;Aventures sur L'île LEGO
+Keywords[it]=LEGO;lego;Isola LEGO
+Keywords[jp]=LEGO;lego;レゴアイランドの大冒険
+Keywords[pt_BR]=LEGO;lego;A Ilha LEGO
```
```suggestion
Keywords=LEGO;lego;LEGO Island
Keywords[da]=LEGO;lego;Panik på LEGO Øen
Keywords[de]=LEGO;lego;Abenteuer auf der LEGO Insel
Keywords[es]=LEGO;lego;La Isla LEGO
Keywords[fr]=LEGO;lego;Aventures sur L'île LEGO
Keywords[it]=LEGO;lego;Isola LEGO
Keywords[ja]=LEGO;lego;レゴアイランドの大冒険
Keywords[ko]=LEGO;lego;레고 아일랜드
Keywords[pt]=LEGO;lego;A Ilha LEGO
Keywords[ru]=LEGO;lego;Остров LEGO
```

That should be all of them

iAmSomeone2 commented:
I've implemented @Damglador's suggestions and fixed the builds. It looks like I'll have to have the save directory default to somewhere in the sandbox, but things are looking good otherwise!

madebr commented on packaging/linux/isledecomp.desktop.in:
```diff
@@ -0,0 +1,35 @@
+
+[Desktop Action configure]
+Name=Configure Settings
+Exec=config
```
```suggestion
Exec=isle-config
```

iAmSomeone2 replied:
The "Exec" key is actually the command used to launch an action. It can be pulled from the $PATH or you can use an absolute path. In the Flatpak, for example, this expands to `/app/bin/config`.

If we really want this name, then I can just update the target name or the install directive in CMake and all builds will use it. Honestly, I think this is the path we should take since `config` is way too general of a name anyway. While Flatpak builds won't suffer from name collision issues, non-sandboxed packages absolutely will.

madebr replied:
We renamed `config` to `isle-config` recently.
https://github.com/isledecomp/isle-portable/commit/73bab247218fee0f6e5e555a08c6a88a33f105e5

iAmSomeone2 replied:
Ah, gotcha. I haven't merged `master` in a few days. I'll get that synced up.

madebr commented on .github/workflows/release.yml:
```diff
@@ -4,6 +4,11 @@ on:
       - master
+      - flatpak
+
+env:
+  APP_ID: org.legoisland.Isle
+  BUILD_DIR: build
```
Let's just hardcode this. `env.BUILD_DIR` is only used twice, and is unlikely to change.
`env.BUILD_DIR` makes the yml unreadable for no gain.

madebr reviewed (commented):
Looks good!
With some small remarks left, I think this one is good to merge.

iAmSomeone2 commented on packaging/linux/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+if(DEFINED ENV{FLATPAK_DEST})
+    set(FLATPAK TRUE)
+else()
+    set(FLATPAK FALSE)
+endif()
```
Yep! Already done in my upcoming commit.

iAmSomeone2 commented:
If this CI run looks good, I'll remove the `flatpak` branch trigger and set this PR as "Ready for review".

AJenbo commented on packaging/linux/isledecomp.metainfo.xml.in:
```diff
@@ -0,0 +1,89 @@
+
+    <keyword>LEGO Island</keyword>
+    <!-- Translations sourced from https://www.pcgamingwiki.com/wiki/Lego_Island#Localizations  -->
+    <keyword xml:lang="fr">Aventures sur L'île LEGO</keyword>
+    <keyword xml:lang="de">Abenteuer auf der LEGO Insel</keyword>
+    <keyword xml:lang="da">Titled Panik på LEGO Øen</keyword>
+    <keyword xml:lang="it">Isola LEGO</keyword>
+    <keyword xml:lang="es">La Isla LEGO</keyword>
+    <keyword xml:lang="jp">レゴアイランドの大冒険</keyword>
+    <keyword xml:lang="pt-BR">A Ilha LEGO</keyword>
```
```suggestion
    <keyword xml:lang="da">Panik på LEGO Øen</keyword>
    <keyword xml:lang="de">Abenteuer auf der LEGO Insel</keyword>
    <keyword xml:lang="es">La Isla LEGO</keyword>
    <keyword xml:lang="fr">Aventures sur L'île LEGO</keyword>
    <keyword xml:lang="it">Isola LEGO</keyword>
    <keyword xml:lang="ja">レゴアイランドの大冒険</keyword>
    <keyword xml:lang="ko">레고 아일랜드</keyword>
    <keyword xml:lang="pt">A Ilha LEGO</keyword>
    <keyword xml:lang="ru">Остров LEGO</keyword>
```

iAmSomeone2 commented on .github/workflows/release.yml:
```diff
@@ -42,8 +42,8 @@ jobs:
-          sudo apt-get update
-          sudo apt-get install -y \
+          sudo apt update
+          sudo apt install -y \
```
Gotcha. It's been a few years since I've used Ubuntu. At the time, the `apt` command looked like it was going to be the catch all for package management, like `dnf` is on RedHat distros.

AJenbo replied:
I had the same misunderstanding

iAmSomeone2 commented:
> Nice work

Thank you! I'm happy to help :)

AJenbo commented:
Next step would be to get it on flathub :)

iAmSomeone2 commented:
That would be awesome!

I'll take some time a bit later to look at their publishing guidelines. I doubt they want us pushing a new version every 20 minutes :sweat_smile:. We could possibly do a "nightly" release at most. Should probably add some screenshots, too.
